src/System.Text.Primitives/System/Text/TextEncoding.cs

```diff
@@ -19,8 +19,68 @@ public enum EncodingName : byte
             //ISO8859_1 = 4,
         }
 
-        public readonly static TextEncoder Utf8 = new Utf8TextEncoding() { Encoding = EncodingName.Utf8 };
-        public readonly static TextEncoder Utf16 = new Utf16TextEncodingLE() { Encoding = EncodingName.Utf16 };
+        #endregion EncodingName enum
+
+        #region Symbol enum
+
+        public enum Symbol : ushort
+        {
+            D0 = 0,
+            D1 = 1,
+            D2 = 2,
+            D3 = 3,
+            D4 = 4,
+            D5 = 5,
+            D6 = 6,
+            D7 = 7,
+            D8 = 8,
+            D9 = 9,
+            DecimalSeparator = 10,
+            GroupSeparator = 11,
+            InfinitySign = 12,
+            MinusSign = 13,
+            PlusSign = 14,
+            NaN = 15,
+            Exponent = 16,
+        }
+
+        #endregion Symbol enum
+
+        #region Static instances
+
+        public readonly static TextEncoder InvariantUtf8 = new Utf8TextEncoding();
+        public readonly static TextEncoder InvariantUtf16 = new Utf16TextEncodingLE();
+
+        #endregion Static instances
+
+        #region Private data
+
+        private readonly EncodingName _encodingName;
+        private readonly byte[][] _symbols;                     // this could be flattened into a single array
+        private readonly ParsingTrie.Node[] _parsingTrie;       // prefix tree used for parsing
+
+        #endregion Private data
+
+        #region Properties
+
+        public EncodingName Encoding => _encodingName;
+
+        public bool IsInvariantUtf8 => this == InvariantUtf8;
+
+        public bool IsInvariantUtf16 => this == InvariantUtf16;
+
+        #endregion Properties
+
+        #region Constructors
+
+        protected TextEncoder(byte[][] symbols, EncodingName encodingName)
```

Comment:
Do we still need EncodingName? Given the context it is used in (that I can see), we could just do a reference check.